include/unifex/sender_concepts.hpp

```diff
@@ -15,98 +15,310 @@
+    } else if constexpr (std::is_base_of_v<sender_base, S>) {
+      return sender_base{};
+    } else {
+      return _no_sender_traits{};
+    }
+  }
+} // namespace detail
+
+template <typename S>
+struct sender_traits : decltype(detail::_select_sender_traits<S>()) {};
+
+template <typename S>
+UNIFEX_CONCEPT //
+  sender =     //
+    move_constructible<remove_cvref_t<S>> &&
+    detail::_has_sender_traits<remove_cvref_t<S>>;
+
+template <typename S>
+UNIFEX_CONCEPT   //
+  typed_sender = //
+    sender<S> && //
+    detail::_has_sender_types<sender_traits<remove_cvref_t<S>>>;
+
+namespace _start_cpo {
+  inline constexpr struct _fn {
+    template(typename Operation)
+      (requires tag_invocable<_fn, Operation&>)
+    auto operator()(Operation& op) const noexcept
+        -> tag_invoke_result_t<_fn, Operation&> {
+      static_assert(
+          is_nothrow_tag_invocable_v<_fn, Operation&>,
+          "start() customisation must be noexcept");
+      return unifex::tag_invoke(_fn{}, op);
+    }
+    template(typename Operation)
+      (requires (!tag_invocable<_fn, Operation&>))
     auto operator()(Operation& op) const noexcept -> decltype(op.start()) {
       static_assert(
           noexcept(op.start()),
           "start() customisation must be noexcept");
       return op.start();
     }
-  };
+  } start{};
 } // namespace _start
-using _start::start;
+using _start_cpo::start;
+
+namespace _connect_cpo {
+  using detail::_can_execute;
+
+  struct _fn;
+
+  template <typename Sender, typename Receiver>
+  using _member_connect_result_t =
+      decltype((UNIFEX_DECLVAL(Sender&&)).connect(
+          UNIFEX_DECLVAL(Receiver&&)));
+  template <typename Sender, typename Receiver>
+  UNIFEX_CONCEPT_FRAGMENT( //
+    _has_member_connect_,  //
+      requires() (         //
+        typename(_member_connect_result_t<Sender, Receiver>)
+      ));
+  template <typename Sender, typename Receiver>
+  UNIFEX_CONCEPT //
+    _with_member_connect = //
+      sender<Sender> &&
+      UNIFEX_FRAGMENT(_connect_cpo::_has_member_connect_, Sender, Receiver);
+  template <typename Sender, typename Receiver>
+  UNIFEX_CONCEPT //
+    _with_tag_invoke = //
+      sender<Sender> && tag_invocable<_fn, Sender, Receiver>;
+  template <typename Executor, typename Receiver>
+  UNIFEX_CONCEPT //
+    _with_execute = //
+      receiver_of<Receiver> && _can_execute<Executor, Receiver>;
 
-namespace _connect {
   inline constexpr struct _fn {
    private:
-    template <bool>
-    struct _impl {
+    template <typename Executor, typename Receiver>
+    struct _as_op {
+      struct type {
+        remove_cvref_t<Executor> e_;
+        remove_cvref_t<Receiver> r_;
+        void start() noexcept try {
+          using _as_invocable =
+            detail::_as_invocable<remove_cvref_t<Receiver>, Executor>;
+          unifex::execute(std::move(e_), _as_invocable{r_});
+        } catch(...) {
+          unifex::set_error(std::move(r_), std::current_exception());
```

Comment:
I know. Suggestions welcome.

Reply:
For the record, I am currently of the opinion that synchronization is worth it here to avoid swallowing the error. Those who care about performance can code to sender/receiver directly.

Reply:
You could also take the view that "those who care about errors can code to sender/receiver directly".

I haven't fully worked through a solution here but I suspect that we may need an additional heap-allocation that contains synchronisation state has joint ownership between the execute() caller and the _as_invocable wrapper.

It may also result in some non-determinism about which execution context the set_error() or set_done() methods would be called from. As it is racy, it could result in either of them being called from either the execute() caller's context, or the executor's context.